docs/auth/backend-auth.md

```diff
@@ -0,0 +1,82 @@
+- `createRequireServerTokenMiddleware`: Creates middleware a JWT bearer token for all requests and verifies that the token is a valid server token (see [service to service Auth](./service-to-service-auth.md)).
+
+> NOTE: `createRequireJwtTokenMiddleware` does not validate the JWT tokens, instead you must specify an express middleware that will be invoked when the token has the specified audience. `createRequireUserTokenMiddleware` and `createRequireServerTokenMiddleware` both validate the tokens.
+
+The middleware created by these factories only verifies that a valid JWT token is provided and thus adds **authentication** to the api endpoint. The middleware cannot check whether the identity enclosed in the token has access to specific resources (**authorisation**). Each plugin backend is responsible for checking the permissions of the identity in the token if required.
+
+To configure all backend routes, except `/auth`, to require a valid user or server token, you can use the following code:
+
+```ts title="packages/backend/src/index.ts"
+// ...
+/* highlight-add-start */
+import {
+  createRequireJwtTokenMiddleware,
+  createRequireServerTokenMiddleware,
+  createRequireUserTokenMiddleware,
+} from '@internal/plugin-auth-node';
+/* highlight-add-end */
+
+// ...
+async function main() {
+  // ...
+  const createEnv = makeCreateEnv(config);
+
+  /* highlight-add-start */
+  const tokenManager = ServerTokenManager.fromConfig(config, {
+    logger: getRootLogger(),
+  });
+  const discovery = SingleHostDiscovery.fromConfig(config);
+  const identity = DefaultIdentityClient.create({
+    discovery,
+  });
+
+  const requireBackstageToken = createRequireJwtTokenMiddleware({
+    backstage: createRequireUserTokenMiddleware(identity),
+    'backstage-server': createRequireServerTokenMiddleware(tokenManager),
+  });
+  /* highlight-add-end */
+
+  const apiRouter = Router();
+
+  /* highlight-add-next-line */
+  // unauthenticated routes
+  apiRouter.use('/auth', await auth(authEnv));
+
+  /* highlight-add-next-line */
+  // authenticated routes
+  apiRouter.use(
+    '/catalog',
+    /* highlight-add-start */ requireBackstageToken,
+    /* highlight-add-end */ await catalog(catalogEnv),
+  );
+  apiRouter.use(
+    '/scaffolder',
+    /* highlight-add-start */ requireBackstageToken,
+    /* highlight-add-end */ await scaffolder(scaffolderEnv),
+  );
+  apiRouter.use(
+    '/techdocs',
+    /* highlight-add-start */ requireBackstageToken.unless({
+      path: /^\/api\/techdocs\/static\/docs\/.+\.css/,
+    }),
+    /* highlight-add-end */ await techdocs(techdocsEnv),
+  );
+  // ...
+}
+```
+
+> Note that for the `/techdocs` route you can use the [unless](https://github.com/jfromaniello/express-unless) method to allow requests to the static CSS files to be unauthenticated. This is because the `techdocs` plugin serves the CSS files from the backend, but the `techdocs` frontend requests them directly from the backend..
```

Comment:
Same thing for fonts and images right? I actually think we need to add cookie-based auth for this case, at least if we need it for images. It feels like fonts and CSS would be alright to expose, but not images.

It's a bit tricky because there's not guarantee that TechDocs is deployed on the same domain as the rest of the plugins, even though this is pretty much always the case in practice. We also don't want to add cookie auth everywhere, since we don't have CSRF protection. What I'm hoping is that we can provide a middleware for the TechDocs backend to use on a route that must always be called before reading TechDocs content, which sets the cookie. For the content we can then use a separate middleware that reads said cookie and uses that for auth.

Reply:
Yeah this was a bit tricky because adding the authentication middleware breaks the TechDocs. I tried the example cookie auth from the [Authenticate API requests](https://github.com/backstage/backstage/blob/master/contrib/docs/tutorials/authenticate-api-requests.md) tutorial but I could not get that to work with the TechDocs. It also passes the token cookie on every request

Reply:
Yep, exactly, that's what I was thinking. Probably something that in the end is common enough for us to provide utilities and a common pattern for